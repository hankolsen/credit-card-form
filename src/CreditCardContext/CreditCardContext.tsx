import React, {
  createContext,
  FunctionComponent,
  useContext,
  useMemo,
  useState,
} from 'react';
import getCreditCardIssuer, {
  CreditCardIssuers,
} from '../getCreditCardIssuer/getCreditCardIssuer';

type CreditCardContextType = {
  isFlipped?: boolean;
  setIsFlipped: (isFlipped: boolean) => void;
  cardNumber: string;
  setCardNumber: (number: string) => void;
  cardName: string;
  setCardName: (name: string) => void;
  cardCvv: string;
  setCardCvv: (cvv: string) => void;
  expirationMonth: string;
  setExpirationMonth: (month: string) => void;
  expirationYear: string;
  setExpirationYear: (year: string) => void;
  formattedCardNumber: string | undefined;
  cardIssuer: string | undefined;
  cardNumberLength: number;
  expirationMontOutput: string;
};

const CreditCardContext =
  createContext<CreditCardContextType | undefined>(undefined);

const CreditCardContextProvider: FunctionComponent = ({ children }) => {
  const [isFlipped, setIsFlipped] = useState<boolean>();
  const [cardNumber, setCardNumber] = useState<string>('');
  const [cardName, setCardName] = useState<string>('');
  const [cardCvv, setCardCvv] = useState<string>('');
  const [expirationMonth, setExpirationMonth] = useState<string>('');
  const [expirationYear, setExpirationYear] = useState<string>('');

  const cardIssuer = getCreditCardIssuer(cardNumber);
  const cardNumberLength =
    cardIssuer === CreditCardIssuers.AMERICANEXPRESS ? 15 : 16;
  let formattedCardNumber = cardNumber
    .padEnd(16, '#')
    .match(/([\d|#]{1,4})/g)
    ?.join(' ');

  if (cardIssuer === CreditCardIssuers.AMERICANEXPRESS) {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const [_, ...amexNumbers] =
      cardNumber
        .padEnd(15, '#')
        .match(/([\d|#]{1,4})([\d|#]{1,6})([\d|#]{1,5})/) || [];
    formattedCardNumber = amexNumbers.join(' ');
  }

  const expirationMontOutput = expirationMonth
    ? expirationMonth.padStart(2, '0')
    : '';

  const value: CreditCardContextType = useMemo(
    () => ({
      isFlipped,
      setIsFlipped,
      cardNumber,
      setCardNumber,
      cardName,
      setCardName,
      cardCvv,
      setCardCvv,
      expirationMonth,
      setExpirationMonth,
      expirationYear,
      setExpirationYear,
      formattedCardNumber,
      cardIssuer,
      cardNumberLength,
      expirationMontOutput,
    }),
    [
      cardCvv,
      cardIssuer,
      cardName,
      cardNumber,
      cardNumberLength,
      expirationMontOutput,
      expirationMonth,
      expirationYear,
      formattedCardNumber,
      isFlipped,
    ],
  );
  return (
    <CreditCardContext.Provider value={value}>
      {children}
    </CreditCardContext.Provider>
  );
};

const useCreditCard = () => {
  const context = useContext(CreditCardContext);
  if (!context) {
    throw new Error(
      'useCreditCard must be used within an CreditCardContextProvider',
    );
  }
  return context;
};

export { CreditCardContextProvider, useCreditCard };
