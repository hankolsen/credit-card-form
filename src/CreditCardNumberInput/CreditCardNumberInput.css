/* Hide those ugly number arrows */
input[type='number']::-webkit-inner-spin-button,
input[type='number']::-webkit-outer-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.card__number--error,
.card__number--error:focus {
  border-color: #b00020;
  background-color: hsl(349, 100%, 90%);
}

.card__number-error-message {
  padding: 0;
  text-align: right;
  margin-top: -28px;
  margin-right: 8px;
  color: #b00020;
  font-size: 0.75rem;
}
