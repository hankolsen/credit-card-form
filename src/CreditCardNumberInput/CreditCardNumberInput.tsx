import React, { ChangeEvent, useState } from 'react';

import './CreditCardNumberInput.css';
import { useCreditCard } from 'CreditCardContext/CreditCardContext';
import validateCreditCard from '../validatedCreditCard/validateCreditCard';

const CreditCardNumberInput = () => {
  const { setCardNumber } = useCreditCard();
  const [error, setError] = useState(false);
  const [numbers, setNumbers] = useState<string>('');
  const wantedLength = 16;
  const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { value } = event.target;
    if (value.length > 16) {
      event.preventDefault();
      return;
    }
    if (value.length > 4) {
      setNumbers(`${value.slice(0, 3)} ${value.slice(4)}`);
    }
    if (value.length === wantedLength) {
      setError(!validateCreditCard(value) ?? false);
    } else {
      setError(false);
    }
    setNumbers(value);
    setCardNumber(value);
  };

  return (
    <>
      <label htmlFor="cardNumber">Card Number</label>
      <input
        name="cardNumber"
        id="cardNumber"
        type="number"
        onChange={handleInputChange}
        maxLength={16}
        value={numbers}
        className={error ? 'card__number--error' : ''}
      />
      {error && <p className="card__number-error-message">That seems wrong</p>}
    </>
  );
};

export default CreditCardNumberInput;
