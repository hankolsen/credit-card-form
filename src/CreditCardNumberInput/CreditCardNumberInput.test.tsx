import React from 'react';
import { render, screen } from '@testing-library/react';
import { CreditCardContextProvider } from 'CreditCardContext/CreditCardContext';
import CreditCardNumberInput from './CreditCardNumberInput';

describe('CreditCardNumberInput test', () => {
  it('should render the CreditCardNumberInput', () => {
    const { container } = render(
      <CreditCardContextProvider>
        <CreditCardNumberInput />
      </CreditCardContextProvider>,
    );
    expect(screen.getByText(/card number/i)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
