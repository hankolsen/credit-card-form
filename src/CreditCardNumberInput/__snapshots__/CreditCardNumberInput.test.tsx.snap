// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreditCardNumberInput test should render the CreditCardNumberInput 1`] = `
<div>
  <label
    for="cardNumber"
  >
    Card Number
  </label>
  <input
    class=""
    id="cardNumber"
    maxlength="16"
    name="cardNumber"
    type="number"
    value=""
  />
</div>
`;
